src/shogun/kernel/TranslationalInvariantKernel.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2016 Soumyajit De
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#include <shogun/lib/common.h>
+#include <shogun/kernel/TranslationalInvariantKernel.h>
+#include <shogun/distance/Distance.h>
+#include <shogun/distance/CustomDistance.h>
+
+using namespace shogun;
+
+CTranslationalInvariantKernel::CTranslationalInvariantKernel()
+	: CKernel(0), m_distance(NULL)
+{
+	register_params();
+}
+
+CTranslationalInvariantKernel::CTranslationalInvariantKernel(int32_t size)
+	: CKernel(size), m_distance(NULL)
+{
+	register_params();
+}
+
+CTranslationalInvariantKernel::CTranslationalInvariantKernel(CFeatures *l, CFeatures *r, int32_t size)
+	: CKernel(l, r, size), m_distance(NULL)
+{
+	init(l, r);
+	register_params();
+}
+
+CTranslationalInvariantKernel::~CTranslationalInvariantKernel()
+{
+	cleanup();
+	SG_UNREF(m_distance);
+}
+
+bool CTranslationalInvariantKernel::init(CFeatures* l, CFeatures* r)
+{
+	CKernel::init(l,r);
+	if (m_distance!=NULL)
```

Comment:
If a subclass (say, GaussianKernel) does not initialize a distance instance, then also it should work, no? For CDistanceKernel, this instance is mandatory.